1810 - Golden Wyrm's ultimate ability

nandastone opened the pull request:
This MR adds Golden Wyrm's ultimate ability as described in https://github.com/FreezingMoon/AncientBeast/issues/1810.

There are outstanding questions in that issue, plus some more raised during development so this is a draft MR. More changes coming soon.

![CleanShot 2021-12-14 at 22 16 33](https://user-images.githubusercontent.com/199204/145996681-159b7416-9064-46b2-b0aa-b57f0c1e6f0e.gif)

Example log:

![CleanShot 2021-12-14 at 21 35 29@2x](https://user-images.githubusercontent.com/199204/145996759-853ddbcb-db18-47ad-a7d1-db5f0023e85d.png)

Cannot be used when ally at full health:

![CleanShot 2021-12-14 at 21 37 21@2x](https://user-images.githubusercontent.com/199204/145996808-86641648-9ad0-4b2f-9691-c9afc7de67fe.png)

Cannot be used when Golden Wyrm under 51 health:

![CleanShot 2021-12-14 at 22 06 05@2x](https://user-images.githubusercontent.com/199204/145996938-9beda42a-bdb3-455f-bd7a-d3bbbe4d4b88.png)

Upgraded ability buffs regrowth:

![CleanShot 2021-12-14 at 22 39 59@2x](https://user-images.githubusercontent.com/199204/146000906-dd987548-cd68-46cd-8542-7dbbee885f33.png)

<a href="https://gitpod.io/#https://github.com/FreezingMoon/AncientBeast/pull/1959"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

nandastone commented on assets/units/data.json:
```diff
@@ -2598,7 +2598,7 @@
 		"id": 33,
 		"name": "Golden Wyrm",
-		"playable": false,
+		"playable": true,
```
@DreadKnight Is there a mechanism for enabling creatures to test?

DreadKnight replied:
Well, this is the mechanism, Golden Wyrm unit has cardboard added. If you do mean something else I'm afraid you'll have to be more specific :D

nandastone replied:
I basically meant is it ok to enable a creature in development?

DreadKnight replied:
@nandastone Oh right, it's a bit of a double edged sword atm tbh. For now hoping you will finish the unit you enabled / working on faster, but player base is low for now. Anyway, I've opened issue #1963 on how to handle this somewhat better.

nandastone commented on src/abilities/Golden-Wyrm.js:
```diff
@@ -233,78 +233,131 @@ export default (G) => {
+				effects. */
+				this.creature.heal(-transferAmount);
+
+				// TODO: Should we show messages/hints for this transfer beyond health lost/gained?
```
@DreadKnight Should there be some log message or hint beyond the loss/gain of health?

![CleanShot 2021-12-14 at 22 16 33](https://user-images.githubusercontent.com/199204/145997474-cdc7fd90-e040-4ec2-8d9b-932b5f1820af.gif)

DreadKnight replied:
The tooltips are fine I'm also pointing out ideal log in another comment in a bit.

nandastone commented on src/abilities/Golden-Wyrm.js:
```diff
@@ -233,78 +233,133 @@ export default (G) => {
+
+					this.creature.addEffect(
+						regrowthBuffEffect,
+						// TODO: What should this say?
```
@DreadKnight I've added a custom log for when the upgraded ability triggers:

<img width="448" alt="CleanShot 2021-12-14 at 22 39 46@2x" src="https://user-images.githubusercontent.com/199204/146001097-4201ec6c-487a-493e-92fa-5a64f7222068.png">

Any feedback on the copy?

DreadKnight replied:
Neat but that's a bit um... cryptic I think.
Should be something more direct I think:
`Golden Wyrm gains 10 regrowth points`
or whatever the default message log is.

nandastone commented on src/abilities/Golden-Wyrm.js:
```diff
@@ -233,78 +233,133 @@ export default (G) => {
+						regrowthBuffEffect,
+						// TODO: What should this say?
+						`%CreatureName${this.creature.id}% is brimming with health`,
+						// TODO: Should we show a hint?
```
@DreadKnight There's currently no hint when the upgraded ability triggers. Should there be one?

DreadKnight replied:
Well, there's a visual "upgrade" animation on the ability icons themselves plus they gain a cooler frame around.
But I guess we could have a message log indicating that an ability has upgraded as well, would make way more sense in some cases while looking around through the log.

`Visible Stigmata has been upgraded`

nandastone replied:
@DreadKnight Sounds good, but can you please creature a separate issue as that change should be applied across the board to other creatures.

DreadKnight replied:
@nandastone Good point. Here it is -> #1962

DreadKnight commented:
> This MR adds Golden Wyrm's ultimate ability as described in #1810.
> 
> There are outstanding questions in that issue, plus some more raised during development so this is a draft MR. More changes coming soon.
> 
> ![CleanShot 2021-12-14 at 22 16 33](https://user-images.githubusercontent.com/199204/145996681-159b7416-9064-46b2-b0aa-b57f0c1e6f0e.gif)
> 
> Example log:
> 
> ![CleanShot 2021-12-14 at 21 35 29@2x](https://user-images.githubusercontent.com/199204/145996759-853ddbcb-db18-47ad-a7d1-db5f0023e85d.png)
> 
Nice, though ideal log would be something like I believe:
`Golden Wyrm transfers 19 health to Gumble`
Provided it fits in just 1 line in most cases, otherwise could use `gives` instead of `transfer` to take even less space.

> Cannot be used when ally at full health:
> 
> ![CleanShot 2021-12-14 at 21 37 21@2x](https://user-images.githubusercontent.com/199204/145996808-86641648-9ad0-4b2f-9691-c9afc7de67fe.png)
> 
Good for now, though this does give me some wild ideas, Quake 3 Arena style: so you would be able to buff allies over their max health, but whatever is outside the pool would constantly drop by 1 each round (even if not taking any damage at all).
Anyway, this is a mechanic to consider for later on, perhaps as an alternative upgrade for the ability as those are planned.

> Cannot be used when Golden Wyrm under 51 health:
> 
> ![CleanShot 2021-12-14 at 22 06 05@2x](https://user-images.githubusercontent.com/199204/145996938-9beda42a-bdb3-455f-bd7a-d3bbbe4d4b88.png)
> 

Good, we don't want him to die just like that or create weird/awkward situations.

> Upgraded ability buffs regrowth:
> 
> ![CleanShot 2021-12-14 at 22 39 59@2x](https://user-images.githubusercontent.com/199204/146000906-dd987548-cd68-46cd-8542-7dbbee885f33.png)
> 

Wingeria! This is basically a "karma" type of ability, give and receive amplified. Plus making a high level unit into a support, woot!

> [![](https://camo.githubusercontent.com/76e60919474807718793857d8eb615e7a50b18b04050577e5a35c19421f260a3/68747470733a2f2f676974706f642e696f2f627574746f6e2f6f70656e2d696e2d676974706f642e737667)](https://gitpod.io/#https://github.com/FreezingMoon/AncientBeast/pull/1959)

DreadKnight commented:
Noticed that ability doesn't work backwards+diagonals as initially expected, but let's actually make it work all around, even sides.
I don't mean AoE ability, but being able to target one spot/unit and heal it, since the Wyrm itself is rather small and humanoid, pretty much in the center tile, so I kinda feel like this should have a generic animation for all sides which will just connect with target via some particle effect, that way we don't complicate things or make them tedious and awkward, it's a game after all.

Also, it should have some energy cost, let's say 25 points, so that it won't break the card template overall and be more balanced.
![Screenshot_20211214_203435](https://user-images.githubusercontent.com/587487/146060343-1097fea2-2852-45c9-9754-6dec5153b03c.jpg)

nandastone commented:
Added a 25 energy cost:

![CleanShot 2021-12-15 at 16 15 59@2x](https://user-images.githubusercontent.com/199204/146133726-57ff666f-6b5a-49c8-89ea-3481d749c301.png)

Changed targeting area to hexes directly adjacent:

![CleanShot 2021-12-15 at 16 14 28@2x](https://user-images.githubusercontent.com/199204/146133785-19931d24-6c08-4106-ab93-97a9c325e448.png)

Updated the log to show "transfers" + regrowth gain:

<img width="447" alt="CleanShot 2021-12-15 at 15 10 12@2x" src="https://user-images.githubusercontent.com/199204/146133839-98647517-c1d6-429b-b30c-fc9c244d093f.png">

nandastone commented on src/abilities/Golden-Wyrm.js:
```diff
@@ -233,78 +233,139 @@ export default (G) => {
+
+					this.creature.addEffect(
+						regrowthBuffEffect,
+						`%CreatureName${this.creature.id}% gains 10 regrowth points`,
```
It'd be great for all alterations to automatically log "x gains x" like drops, but that's a larger change for another PR.

nandastone commented on src/abilities/Golden-Wyrm.js:
```diff
@@ -233,78 +233,139 @@ export default (G) => {
+					[1, 0, 0, 0, 1],
+					[1, 1, 1, 1, 0],
+				];
+				const xOffset = this.creature.y % 2 === 0 ? 0 : 1;
```
Can't help but feel there is a helper function that selected hexes while abstracting away the need to check odd/even rows, but was unable to find/understand it.

The hexes are one of the most confusing parts of the codebase to me, still working it out.

DreadKnight commented:
Seems to be working good overall, well done once again! 🥳 
Found some low priority visual issue #1965 for v0.5 milestone.

DreadKnight commented:
> Added a 25 energy cost:
> 
> ![CleanShot 2021-12-15 at 16 15 59@2x](https://user-images.githubusercontent.com/199204/146133726-57ff666f-6b5a-49c8-89ea-3481d749c301.png)
> 

I see, so cost was initially 50 actually but not implemented :D will keep that in mind xD

> Changed targeting area to hexes directly adjacent:
> 
> ![CleanShot 2021-12-15 at 16 14 28@2x](https://user-images.githubusercontent.com/199204/146133785-19931d24-6c08-4106-ab93-97a9c325e448.png)
> 

Nice, was interested you haven't skipped the sides, since the Golden Wyrm can reach those rather easy if extending hand out.

> Updated the log to show "transfers" + regrowth gain:
> 
> <img alt="CleanShot 2021-12-15 at 15 10 12@2x" width="447" src="https://user-images.githubusercontent.com/199204/146133839-98647517-c1d6-429b-b30c-fc9c244d093f.png">

Replaced with "gives" to avoid that double row overflow visual issue in order to keep things nicer visually overall 😎
